src/policy/concrete.rs

```
@@ -137,30 +143,109 @@ impl fmt::Display for PolicyError {
+        Ok(taptree)
     }
-    /// Extract the Taproot internal_key from policy tree.
+
+    /// Extract the internal_key from policy tree.
     #[cfg(feature = "compiler")]
-    fn extract_key(policy: &Self, unspendable_key: Option<Pk>) -> Result<(Pk, &Policy<Pk>), Error> {
-        match unspendable_key {
-            Some(key) => Ok((key, policy)),
-            None => Err(errstr("No internal key found")),
+    fn extract_key(policy: &Self, unspendable_key: Option<Pk>) -> Result<(Pk, Policy<Pk>), Error> {
+        let semantic_policy = policy.lift()?;
+        let concrete_keys = policy.keys().into_iter().collect::<HashSet<_>>();
+        let mut internal_key: Option<Pk> = None;
+
+        for key in concrete_keys.iter() {
+            if semantic_policy
+                .clone()
+                .satisfy_constraint(&Semantic::KeyHash(key.to_pubkeyhash()), true)
+                == Semantic::Trivial
+            {
+                internal_key = Some((*key).clone());
+                break;
+            }
+        }
+
+        match (internal_key, unspendable_key) {
+            (Some(key), _) => Ok((key.clone(), policy.translate_unsatisfiable_pk(&key))),
+            (_, Some(key)) => Ok((key, policy.clone())),
+            _ => Err(errstr("No viable internal key found.")),
         }
     }
 
     /// Compile the [`Tr`] descriptor into optimized [`TapTree`] implementation
+    // TODO: We might require other compile errors for Taproot. Will discuss and update.
     #[cfg(feature = "compiler")]
     pub fn compile_tr(&self, unspendable_key: Option<Pk>) -> Result<Descriptor<Pk>, Error> {
         let (internal_key, policy) = Self::extract_key(self, unspendable_key).unwrap();
-        let tree = Descriptor::new_tr(
-            internal_key,
-            Some(Self::compile_huffman_taptree(policy).unwrap()),
-        )
-        .unwrap();
+        let tree =
+            Descriptor::new_tr(internal_key, Some(policy.compile_tr_policy().unwrap())).unwrap();
```

This should either propagate the error or explain why the unwrap is safe

---

Refactored code to propagate errors in [b36c116](https://github.com/rust-bitcoin/rust-miniscript/pull/291/commits/b36c116babe70817942f2bcd95b97a6d00289ff7)